File: Prob3.py
# // Time Complexity : O(m+n)
# // Space Complexity : O(1)
# // Did this code successfully run on Leetcode : Yes
# // Three line explanation of solution in plain english
# Here, we know each row and col is sorted. Upon close observation we can see that from top-right corner and bottom-left corner elements increase and decrease row and column vise (or vice versa). Essentially, we can make a decision on how to move.
# Thus, start at one of the 2 corners, if curr element>target, go in the direction of smaller elements, if curr element<target, go in the direction of bigger elements.

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        #Method 1 - Starting from Top-Right corner
        # m=len(matrix)
        # n=len(matrix[0])
        # r=0
        # c=n-1
        # while 0<=r<m and 0<=c<n:
        #     if matrix[r][c]==target:
        #         return True
        #     elif matrix[r][c]<target: #Current element is smaller than target, so go in the direction of bigger elements -> next row
        #         r+=1
        #     else:
        #         c-=1  #Current element is bigger than target, so go in the direction of smaller elements -> next col
        #Method 2 - Starting from Bottom-Left corner
        m=len(matrix)
        n=len(matrix[0])
        r=m-1
        c=0
        while 0<=r<m and 0<=c<n:
            if matrix[r][c]==target:
                return True
            elif matrix[r][c]>target:   #Current element is bigger than target, so go in the direction of smaller elements -> next row 
                r-=1
            else:               #Current element is smaller than target, so go in the direction of bigger elements -> next col
                c+=1 

File: prob3.cpp
240. Search a 2D Matrix II

// Time Complexity :O(M+N)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :YES
// Three line explanation of solution in plain english

// placing pointer at a place where we can move pointer based on certain condition.Here,
// I will place at lower left corner,as the rows and columns are sorted ,I will compare target with the current 
// value if it is less I will top side else I will move right side

// Your code here along with comments explaining your approach



class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int j=0;
        int i= matrix.size()-1;
// here pointer can be started from i=matrix[0].size()-1 and j=0 as well ,pointer 
// can be placed such that on conditions we can move pointer in a decisive manner.
        while(i>=0 && j<matrix[0].size())
        {
            if(matrix[i][j]==target)
            return true;
            else if(target<matrix[i][j])
            {
                i--; // moving up
            }
            else
            {
                j++; // moving right
            }
        }
        return false;
    }
};